TSS-1580: Add enhanced support for wpf property bindings and expression properties

jkj3shape opened the pull request:
# Issues / Challenges

`Properties` would be not be `Properties` after obfuscation.
`Properties` which was `internal` or inherited and in relation to any implementation accessible to WPI binding would not for all cases be `Skipped`. This was mostly `internal` in general and `Properties` from base classes where heir(s) implement `INotifyPropertyChanged` `interface`.
Enum's would in general also be affected by this regarding being accessible by WPF bindings from heir(s).
(Solution to skip all enum's in settings file has been added to the settings file configuration setup `MSBuild` `Task` to avoid any use of enum's in serialization to fail along with solving know issues with enum's in general in Obfuscar.

# Solution

Skip properties which are internal and not just public.
Skip properties which are public or internal and either by enum types or by property directly if either type or property is related to any type implementing `INotifyPropertyChanged` implementation.
Fix generic types count to avoid exceptions when logging type names.
Avoid logging warnings for methods skipped by enabling `KeepPublicAPI` setting.

# Testing

Unit tests for both `ImplementsInterface` and `HeirsImplementsInterface` have been added both for types and properties.
Unit tests for internal properties have been updated for skipping enum's and properties.

Commit:
Fix convert member names from Type<T>.Member to Type`1.Member form

Commit:
Disregard public properties for types which has heirs implementing INotifyPropertyChanged

Commit:
Add tests for TypeDefinition ImplementInterfaces extension method

Commit:
Exclude all public enums also with HeirsImplementsInterface

Commit:
Refactor HeirsImplementsInterface and dd tests for it and ImplementsInterface

Commit:
Skip not only public but all accessible properties related to INotifyPropertyChanged implementations

Commit:
Add only enum types names to XAML type names to skip

Extend only types in XAML with enum type names not only for public but for all accessible types

Commit:
Fix and skip internal properties in relation to implementation(s) of INotifyPropertyChanged

Commit:
Skip showing warning for methods skipped by KeepPublicApi option

Commit:
Skip internal fields and properties of types known in XAML

Commit:
Avoid replacing properties with getter/setter methods and instead rename

Many problems can occur and have been observed when replacing properties with methods when used Expressions

Commit:
Correct the generic types calculation to include all nested generic types

mv3shape commented on Obfuscar/Helpers/MethodDefinitionExtensions.cs:
```diff
@@ -8,5 +9,37 @@ public static bool IsPublic(this MethodDefinition method)
+                       method.IsPublic ||               //public
+                       method.IsAssembly ||             //internal
+                       //method.IsFamily ||             //protected
+                       method.IsFamilyAndAssembly ||
```
Did you check in the mapping file or with a test if this gets protected members excluded?

jkj3shape replied:
Done through unit tests

mv3shape replied:
So protected members are not excluded?

mv3shape commented on Obfuscar/Helpers/TypeDefinitionExtensions.cs:
```diff
@@ -123,5 +124,43 @@ public static bool ImplementsInterface(this TypeDefinition type, string interfac
             return false;
         }
+
+        public static bool HeirsImplementsInterface(this TypeDefinition type, string interfaceName, IEnumerable<AssemblyInfo> assemblies)
```
Please add a comment to the method

mv3shape commented on Obfuscar/Obfuscator.cs:
```diff
@@ -148,47 +148,49 @@ public void RunRules()
+            void AddToNamesInXamlIfRequired(TypeReference type)
+            {
+                var typeDefinition = type.Resolve();
+                if (typeDefinition != null && typeDefinition.IsEnum)
```
We should actually consider including more than enums, and e.g. dependency property classes should also be excluded

jkj3shape replied:
Currently we are trying to handle all in this collect method.
New implementation does not change this but adds a method more to detect all types in relation to an implementation of `INotifyPropertyChange`.

I actually wanted to remove the enum types not related to the BAML collected types and only use that for properties as it probably should have been done like that in the first place, but I do not want to break any backward combability.

Also worth mentioning: We only use types from BAML extraction - we could look into `PropertyRecord`s to combine that with public/internal properties in types in relation to an implementation of `INotifyPropertyChange`. This could probably give us a few more percent from things not currently used in bindings but do support it.

This I also chose not to pursue as we have lots of manual notified properties which use strings or other logic which could be broken by such changes.

Lots of room for improvement for the future but lets take it in steps.

mv3shape commented on Obfuscar/Obfuscator.cs:
```diff
@@ -1301,10 +1303,16 @@ public void RenameMethods()
-                if (Project.Settings.AbortOnInconsistentState)
-                    throw new ObfuscarException(message.ToString());
+                //if not skipped by KeepPublicApi option in configuration
+                if (!@group.Methods.Select(m => Mapping.GetMethod(m))
```
Two whole point of this warning/exception is to notify if e.g. a member is marked for exclusion in one implementation of an interface and not in another implementation, so I don't think this check is correct.

jkj3shape replied:
It has been removed and I added a setting instead to skip logging warnings about virtual methods where obfuscation has recovered from an inconsistent state.

mv3shape commented on ObfuscarTest/FluentAssertions/Mono.Cecil/EventDefinitionAssertions.cs:
```diff
@@ -0,0 +1,149 @@
+﻿using System.Collections.Generic;
```
What are these files? Don't they exist in a nuget package?

jkj3shape replied:
All these files are fluent assertion additions for making test assertions DRY and clean.

mv3shape commented on Obfuscar/MethodKey.cs:
```diff
@@ -180,10 +180,17 @@ static string ToGenerelGenericMemberName(string memberName)
                         level--;
+                        
+                        //if still within generic or nested generic type listings
+                        if (level >= 0)
```
Could you add some tests for this?

jkj3shape replied:
Did some more debugging and found out it should only increment on first level of generic types.
Fixed it and added some basic tests to assure it works with non generic and generic type names.

Commit:
Avoid logging "Inconsistent virtual method obfuscation state detected" when skipped by configuration or assembly attribute

Skip logging warning If all methods in group are skipped either by KeepPublicAPI configuration or Obfuscate/Obfuscation attribute

Commit:
Add option LogRecoveredInconsistentState to enable/disable warning logging of recovered virtual method inconsistent states

Commit:
Optimize and simplify skipping of types related to any implementation of INotifyPropertyChange

Commit:
Add code XML documentation

Commit:
Fix ToGeneralGenericMemberName to not increment for nested generics

Commit:
Revert addition of redundant property declaring type filter

Commit:
Refactor to extract method to collect all none <Module> assemblies types

Commit:
Extract all FluentAssertions specialized extensions to own project

mv3shape commented on Obfuscar/AssemblyInfo.cs:
```diff
@@ -1224,18 +1227,12 @@ public bool ShouldSkipStringHiding(MethodKey method, InheritMap map, bool projec
-            if (typeInXaml && prop.Property.IsPublic())
+            if (declaringTypeRelatedToNpcOrInXaml && prop.Property.IsPublicOrInternal())
             {
                 message = "filtered by BAML";
```
Could we write "filtered by BAML or INotifyPropertyChanged"?

jkj3shape replied:
3541229f10b1499fa2b3cf3bd89ea4c4ef0c2c72

mv3shape commented on Obfuscar/Helpers/TypeDefinitionExtensions.cs:
```diff
@@ -123,5 +127,66 @@ public static bool ImplementsInterface(this TypeDefinition type, string interfac
+
+            return types
+                .Where(type => type.ImplementsInterface(interfaceName))
+                .ToHashSet();
```
Why is it converted to a hash set? As far as I can see it is never used as such. Is it to remove duplicates?

jkj3shape replied:
Good point --> it was required earlier before last optimization and refactoring, but now its no longer needed.

jkj3shape replied:
4ba3f1ed2d6916e31f5dbab1bdb6d32a874888fe

mv3shape commented on Obfuscar/Obfuscator.cs:
```diff
@@ -148,47 +154,83 @@ public void RunRules()
+            var allEnumTypeNamesRelatedToNpc = CollectAllTypes()
+                .SelectMany(type => type.Properties.Where(PublicOrInternalEnumPropertyFilter))
+                .Where(property => allTypeNamesRelatedToNpc.Contains(property.DeclaringType.FullName))
+                .Select(property => property.PropertyType.Resolve().FullName);
```
In the previous version there was a check if Resolve() returns null. Can that happen? Do we still need that check?

jkj3shape replied:
Nice observed but no it is no longer needed as it is done in the `PublicOrInternalEnumPropertyFilter` check.
Hmm It will cost nothing so I will add it to make sure we do not run into issues with it in the future if the check is left out of the `PublicOrInternalEnumPropertyFilter` check.

jkj3shape replied:
1eeadd344f7ddc44ef90cd6ce944d11c9cce88e6

mv3shape commented on ObfuscarTest.FluentAssertions/Mono.Cecil/EventDefinitionAssertions.cs:
```diff
@@ -0,0 +1,149 @@
+                .ForCondition(map.GetEvent(new EventKey(Subject))?.Status == status)
+                .FailWith("Expected {context:" + Identifier + "} " +
+                          "for event {0} to have obfuscation status {1} but found status {status}.",
+                    Subject, status);
```
How does this format string work? Will {status} be magically replaced by something`?

jkj3shape replied:
```
            Execute.Assertion.AddReportable("status",
                () => Subject != null
                    ? map.GetEvent(new EventKey(Subject))?.Status.ToString()
                    : null);
```

Commit:
Update property skipped message if skipped due to BAML or INotifyPropertyChanged

Commit:
Avoid unnecessary ToHashSet construction for type definition interface detection

Commit:
Add null filter explicit for enum property type name resolves
